### DISCUSSION TITLE
0.3.0

### DISCUSSION THREAD

**Discussion by schickling** [contentlayerdev/contentlayer#356] [Announcements]
ℹ️ **[TLDR]** New experimental source and required peer dependency update.

## ⚠️ Breaking Change: Updated esbuild Dependency

`0.3.0` requires use of [esbuild `0.17.0`](https://github.com/evanw/esbuild/releases/tag/v0.17.0). You may need to update peer dependencies if experiencing installation issues.

## ✨ New Source: Remote Files [experimental]

While still focused on content coming from files, you can begin to explore loading content from files not located in your repository.

This works by syncing content from a remote location into your local workspace, and then behaves similarly to the files source. Contentlayer provides the hook (via a `syncFiles` property) for syncing the files, but you must write the code that pulls the files in.

[Here is simple example](https://github.com/contentlayerdev/contentlayer/tree/main/examples/node-script-remote-content) with a remote Git repo and [documentation](https://www.contentlayer.dev/docs/sources/remote-files).

```ts
import { makeSource } from 'contentlayer/source-remote-files'

export default makeSource({
  syncFiles: () => syncContentFromGit(),
  contentDirPath: 'remote-content',
  documentTypes: [Post],
  disableImportAliasWarning: true,
})

const syncContentFromGit = async () => {
  const syncRun = async () => {
    const repoAlreadyCloned = false
    if (repoAlreadyCloned) {
      // TODO `git clone` the repo
    } else {
      // TODO `git pull` the repo
    }
  }

  let wasCancelled = false
  let syncInterval

  const syncLoop = async () => {
    await syncRun()

    if (wasCancelled) return

    syncInterval = setTimeout(syncLoop, 1000 * 60)
  }

  syncLoop()

  return () => {
    wasCancelled = true
    clearTimeout(syncInterval)
  }
}
```

## ✨ New helper functions: `defineComputedFields` & `defineFields`

You can now use a `defineComputedFields` function to leverage the document type, including its static fields. Here's an example:

```ts
import { defineDocumentType, defineComputedFields } from 'contentlayer/source-files'

const computedFields = defineComputedFields<'Post'>({
  upperTitle: {
    type: 'string',
    resolve: (doc) => doc.title.toUpperCase(),
  },
})

const Post = defineDocumentType(() => ({
  name: 'Post',
  filePathPattern: `**/*.md`,
  fields: {
    // ...
  },
  computedFields,
}))
```

## Other Improvements

- `mdxOptions` now always applies default Contentlayer remark plugins.
- Fixed a bug that avoids the issue demonstrated in #306.
- Upgraded dependencies.

<hr /><em>This discussion was created from the release <a href='https://github.com/contentlayerdev/contentlayer/releases/tag/v0.3.0'>0.3.0</a>.</em>

**Comment by airtonix** [comment]
Will `defineComputedFields` solve the following problems? 


- https://github.com/contentlayerdev/contentlayer/issues/358
- https://github.com/contentlayerdev/contentlayer/issues/344
- https://github.com/contentlayerdev/contentlayer/issues/277
- https://github.com/contentlayerdev/contentlayer/issues/182
- https://github.com/contentlayerdev/contentlayer/issues/149
